org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java

```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
+        .build();
+
+    DoStructureMigrationContext context = BEANS.get(DoStructureMigrationContext.class)
+        .putGlobal(BEANS.get(DoValueMigrationIdsContextData.class)
+            .withAppliedValueMigrationIds(Collections.emptySet())); // run all value migrations
+
+    // migration run 1
+    DoStructureMigratorResult<RoomFixtureDo> result = s_migrator.migrateDataObject(context, original, RoomFixtureDo.class);
+
+    assertTrue(result.isChanged());
+
+    IDoEntity expected = BEANS.get(RoomFixtureDo.class)
+        .withName("example")
+        .withAreaInSquareMeter(20); // increased by 10 through RoomSizeValueMigrationHandler
+
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+
+    // migration run 2: data object is changed again, due to non-idempotent value migration
+    result = s_migrator.migrateDataObject(context, result.getDataObject(), RoomFixtureDo.class);
+
+    assertTrue(result.isChanged());
+
+    expected = BEANS.get(RoomFixtureDo.class)
+        .withName("example")
+        .withAreaInSquareMeter(30); // increased by 10 through RoomSizeValueMigrationHandler
+
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+  }
+
+  /**
+   * Tests StringId value rename with {@link HouseTypeValueMigrationHandler}.<br>
+   * Migration handler matches specific ID type.
+   */
+  @Test
+  public void testRenameStringIdValue() {
+    // raw DoEntity data object
+    IDoEntity original = BEANS.get(DoEntityBuilder.class)
+        .put("_type", "charlieFixture.HouseFixture")
+        .put("_typeVersion", CharlieFixture_2.VERSION.unwrap()) // latest version, no structure migration will be applied
+        .put("name", "example")
+        .put("houseType", "house") // will be migrated by HouseTypeValueMigrationHandler
+        .putList("rooms", Collections.emptyList())
+        .build();
+
+    // migration run 1
+    DoStructureMigratorResult<HouseFixtureDo> result = s_migrator.migrateDataObject(s_migrationContext, original, HouseFixtureDo.class);
+
+    assertTrue(result.isChanged());
+
+    IDoEntity expected = BEANS.get(HouseFixtureDo.class)
+        .withName("example")
+        .withHouseType(HouseTypeFixtureStringId.of("detached-house"))
+        .withRooms(Collections.emptyList());
+
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+
+    // migration run 2: re-apply the same (idempotent) value migration, no changes should occur
+    result = s_migrator.migrateDataObject(s_migrationContext, result.getDataObject(), HouseFixtureDo.class);
+
+    assertFalse(result.isChanged());
+
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+  }
+
+  /**
+   * Tests recursive value migrations with {@link HouseValueMigrationHandler} and
+   * {@link RoomTypeValueMigrationHandler}.<br>
+   * Migration handler matches and replaces whole data object (RoomFixtureDo)
+   */
+  @Test
+  public void testReplaceDataObject() {
```

Comment:
Combine a test that adds a value via structure migration that is migrated via value migration